README.md
# Mapserver

## 更新日志

[更新日志](./Changelog.md)

## 构建镜像

执行脚本：

```bash
chmod +x ./build.sh
./build.sh
```

## 导出镜像至文件

指定存储位置和文件名称，将构建的镜像打包成 tar 文件

```bash
docker save zxht/wmts-mapserver:1.1.0 > ./wmts-mapserver@1.1.0.tar
```

## 导入镜像至服务器

将上一步骤的导出的 `wmts-mapserver@1.1.0.tar` 文件拷贝到服务器，使用下面命令还原成 docker images。

```bash
docker load -i ./wmts-mapserver@1.1.0.tar
```

### 配置 docker-compose.yml 启动

```yml
google-wmts:
  container_name: mapserver
  image: zxht/wmts-mapserver:1.1.0
  environment:
    - PORT=5555
    - HOST=127.0.0.1
  ports:
    - 5555:5555
  privileged: true
```

注意我们在上面的 docker-compose.yml 中配置了 `HOST` 和 `POST` 两个环节变量。

Changelog.md
# 变更历史

## 1.2.0

*2018-06-24*

### 新增

- 增加环境变量：
  - `TILE_TYPE`, 默认值  `default`, 可选值：`GOOGLE`
    当 设置 TILE_TYPE=GOOGLE 时，支持如下结构的 TILE：
    ```
    /L11/R00000328
    - C00000698.png
    - C00000699.png
    - C0000069a.png
    - C0000069b.png
    - C0000069c.png
    - C0000069d.png
    - C0000069e.png
    - C0000069e.png
    ```
     

## 1.1.0

*2018-05-21*

### 新增

- 增加环境变量：
  - `HOST`， 用于设置服务主机地址，默认值 `127.0.0.1`；
  - `POST`， 用于设置服务端口号，默认值 `5555`；

walkDir.py
#coding=utf-8
#Filename:walkDir.py
import os,  os.path
import sqlite3
import os
import string,  cStringIO
from PIL import Image

try:
    from sqlite import encode,  decode
except ImportError:
    import base64
    sqlite3.encode = base64.encodestring
    sqlite3.decode = base64.decodestring
else:
    sqlite3.encode = encode
    sqlite3.decode = decode

def init_db(path,  name):
    '''
    @param path:新建数据库路径
    @param name: 数据库名称
    @return: 返回新建的数据库
    '''
    conn = sqlite3.connect(path + os.sep + name + ".sqlite")
    c = conn.cursor()
    c.execute('''select count(*) from sqlite_master where type='table' and tbl_name='layer' ''')
    if c.fetchall()[0][0] == 0:
        c.execute('''create table layer (z, x, y, data)''')
    conn.commit()
    c.close()
    return conn

def visitDir(img_path, db_path, db_name):
    conn = init_db(db_path, db_name)
    c = conn.cursor()
    n = 0
    for root, dirs, files in os.walk(img_path):   #遍历文件夹目录，几百万个内存是否会占满？？？效率问题
        for filepath in files:
            try:
                tilepath = os.path.join(root, filepath)
                xyz = tilepath.split(os.sep)
                z = int(xyz[-3][1:])
                y = int(xyz[-2][1:9], 16)
                x = int(xyz[-1][1:9], 16)
                # print tilepath

                inn = cStringIO.StringIO()
                Image.open(tilepath).save(inn, "png")

                sql = "insert into layer (z, x, y, data) values (%d, %d, %d, ?)"%(z, x, y)
                c.execute(sql,  (sqlite3.Binary(inn.getvalue()), ))
                n = n + 1
                if n % 5000==0:
                    conn.commit()
                    print n
            except Exception, Argument:
                print Argument
                continue
        conn.commit()
    conn.commit()
    c.close()
    print n
    print u'地图瓦片已全部成功存入数据库'



if __name__ == '__main__':
    # img_path = r'C:\jc'
    # db_path = r'c:'
    # db_name = 'jincheng4'
    visitDir('/Users/liurongtao/Documents/Traffic/GoogleWMTS/tiles', '/Users/liurongtao/Documents/Traffic/GoogleWMTS', 'huizhou')

    # 数据库中创建索引的代码为：create xyz_index on layer(x, y, z)

TileCache.py
#!/usr/bin/env python
#coding=utf-8

import tornado.web
import sqlite3
    
import cStringIO
from PIL import Image

img = Image.new("RGBA",(256,256),(128,0,0,0))
tmp = cStringIO.StringIO()
img.save(tmp,'PNG')
blank = tmp.getvalue()

class SQLiteTile:
    #缓存
    CACHE = {}
    #缓存最大值
    CACHE_MAXIMUM = 10000
    #查询缓冲大小,请求某张图片时将连带下载周边的图片，以减少数据库查询次数
    BUF_SIZE = 1
    #图片标识符格式模板 "layer-x-y-z"
    TILE_ID_TMPL = "%s-%d-%d-%d"
    #图片查询SQL模板
    QUERY_TMPL = "select x,y,z,data from %s where x>%d and x< %d and y>%d and y<%d and z=%d"
    #数据库中没有相应记录的图片uid记录集
    NO_STORE = {}    

    def __init__(self,p):
        self.path = p

    def open(self):
        self.connect = sqlite3.connect(self.path)
        self.cursor = self.connect.cursor()

    def close(self):
        self.connect.close()

    def guid(self,layer,x,y,z):
        return self.TILE_ID_TMPL%(layer,x,y,z)

    def get(self,layer,x,y,z):
        guid = self.guid(layer,x,y,z)

        if guid in self.NO_STORE:
            print "tile %s not exsit"%guid
            return None

        data = self.CACHE.get(guid)

        if data is not None:
            return data

        sql = self.QUERY_TMPL%(layer,
            (x - self.BUF_SIZE),
            (x + self.BUF_SIZE),
            (y - self.BUF_SIZE),
            (y + self.BUF_SIZE),
            z)

        result = self.cursor.execute(sql).fetchall()
        d = dict([(self.guid(layer,*rec[:3]),rec[-1]) for rec in result])
        self.CACHE.update(d)

        data = self.CACHE.get(guid)

        if data is None:
            self.NO_STORE.update({guid:None})

        return data

class PGISTileHandler(tornado.web.RequestHandler):
    def initialize(self,st):
        self.st = st

    def get(self):
        x,y = int(self.get_argument("Col")),int(self.get_argument("Row"))
        z = int(self.get_argument("Zoom"))+int(self.get_argument("ZoomOffset"))
        tile = self.st.get("jiangsu",x,y,z)
        if tile is None:
            self.write(blank)
            return
        self.set_header("Content-Type", "image/png")
        self.write(str(tile))
        return

class TMSTileHandler(tornado.web.RequestHandler):
    def initialize(self,st):
        self.st = st

    def get(self, x, y, z):
        tile = self.st.get("layer", int(x), int(y), int(z))
        self.set_header("Content-Type", "image/png")
        if tile is None:
            self.write(blank)
            return
        self.write(str(tile))
        return


if __name__ == '__main__':
    import tornado.ioloop
    import tornado.web
    from tornado import httpserver

    st = SQLiteTile("huizhou.sqlite")
    st.open()
    # dta = st.get("jiangsu",481,126,11)
    # print str(dta)
    handlers = [
        (r"/tiles",PGISTileHandler,{"st":st},),
        (r"/TMS/(\w+)/(\w+)/(\w+)",TMSTileHandler,{"st":st},)
    ]

    application = tornado.web.Application(handlers)
    
    http_server = httpserver.HTTPServer(application)
    http_server.listen(9001)
    tornado.ioloop.IOLoop.instance().start()

myDownMaps.py
#coding=utf-8
#Filename:Dmc_cmd.py
import cmath, urllib, os, math, random, sys, time, threading
import Queue
import datetime

from walkDir import *

class Cdt:
    def __str__(self):
        return '{x = %s, y = %s}'%(self.x, self.y)

def createName(num, rc):
    '''创建文件夹/文件名称'''
    temp = '00000000'+str(hex(int(num)))[2:]
    return rc + temp[::-1][0:8][::-1]

def getPixelFromCdt(x, y, z):
    '''根据经纬度坐标以及缩放等级获取像素坐标'''
    pixel = Cdt()
    sinLatitude = cmath.sin(y * cmath.pi / 180)
    pixel.x = ((x + 180) / 360) * 256 * (2**z)
    temp = cmath.log((1+sinLatitude)/(1-sinLatitude))
    pixel.y = abs((0.5 - temp / (4 * cmath.pi))*256*(2**z))
    return pixel

def getTileFromPixel(pixel):
    #'''根据像素坐标获取切片'''
    tile = Cdt()
    tile.x = math.floor(pixel.x / 256)
    tile.y = math.floor(pixel.y / 256)
    return tile

def getTileFromCdt(x, y, z):
    '''根据经纬度坐标以及缩放等级获取切片'''
    return getTileFromPixel(getPixelFromCdt(x, y, z))

def createCacheStruc(extent, lvRange, cacheDir):
    '''创建缓存目录结构及计算tile'''
    global tempTask
    tempTask = Queue.Queue()    #事先把队列创建好之后再创建并启动线程，保证qsize正常

    print u'创建Cache目录及计算tile数目...'

    for lv in lvRange:

        if lv < 10:
            lvName = 'L0' + str(lv)
        else:
            lvName = 'L'+str(lv) 

        startTile = getTileFromCdt(extent[0],extent[1],lv)
        endTile = getTileFromCdt(extent[2],extent[3],lv)
        xRange = range(int(startTile.x),int(endTile.x))
        yRange = range(int(startTile.y),int(endTile.y))
        for row in yRange:
            rowName = cacheDir +os.sep + lvName + os.sep + createName(row,'R') 
            if not os.path.exists(rowName):
                os.makedirs(rowName)  

            for col in xRange:
                tempTask.put('%s,%s,%s'%(lv, row, col))

def createRemoteUrl(x, y, z):
    '''创建远程tile地址'''
    port = str(random.randint(0, 3))
    url = 'http://mt%s.google.cn/vt/lyrs=m@169000000&hl=zh-CN&gl=cn&x=%s&y=%s&z=%s&s='%(port, x, y, z)
    print url
    return url

def createLocalFile(x, y, z, cacheDir):
    '''创建缓存本地路径
    计算<等级目录>名称'''

    if int(z) < 10:
        l = 'L0' + str(z)  #x,y,z为字符串类型
    else:
        l = 'L' + str(z)

    #计算<行目录>和<(列)图片>名称
    r = createName(y, 'R')
    c = createName(x, 'C')
    #拼装本地路径
    return cacheDir + os.sep + l + os.sep + r + os.sep + c + '.png'

def downloadTile(remoteFile, localFile):
    '''下载远程文件到本地'''
    urllib.urlretrieve(remoteFile, localFile)

def loadTask(task, threadNum):
    global failureTask
    failureTask = []
    global tasksize
    tasksize = tempTask.qsize()

    print u'待下载Tile总计:%s,下载线程数:%s'%(tasksize, threadNum)
    print u'开始下载'
    print u'下载中...'

    for i in range(threadNum):
        tempTask.put(None)
    for i in range(threadNum):
        Download(tempTask).start() #生成多个线程并启动
   
class Download(threading.Thread):
    progress = 0
    # sucessCount = 0
    failureCount = 0

    def __init__(self, queue):
        threading.Thread.__init__(self)
        self.lock = threading.RLock()
        self.queue = queue

    def run(self):
        while 1:
            item = self.queue.get()
            if item is None:
                self.queue.task_done()
                break
            valueAry = item.split(",")
            lv = valueAry[0]
            row = valueAry[1]
            col = valueAry[2]
            remoteFile = createRemoteUrl(col, row, lv)
            localFile = createLocalFile(col, row, lv, cacheDir)
            try:
                downloadTile(remoteFile, localFile)
            except:
                logStr = '%s,%s,%s'%(lv, row, col)
                f = file(cacheDir + os.sep + 'error.log','a')#在日志文件中打印失败记录
                f.write(logStr+'\n')
                f.close()

                self.lock.acquire()
                Download.failureCount += 1
                print u'下载失败数量：%s'%Download.failureCount #记录下载失败数量
                self.lock.release()

                failureTask.append(logStr)

            
            self.lock.acquire()
            self.queue.task_done()
            Download.progress = Download.progress + 1 #记录下载总数量，包括失败数量
            if Download.progress%100 == 0:
                print u'下载进度为:%s/%s'%(Download.progress, tasksize)
            self.lock.release()

if __name__ == '__main__':
    extent = '114.350570 23.049234 114.501218 23.149616'
    maxLv = '18'
    minLv = '8'

    global extAry, lvRange, cacheDir, threadNum
    ext = [float(i) for i in extent.split(' ')]
    extAry = [ext[0], ext[3], ext[2], ext[1]]          #区域范围
    lvRange = range(int(minLv), int(maxLv) + 1)      #等级范围
    cacheDir = '/Users/liurongtao/Documents/Traffic/GoogleWMTS/tiles'      #下载目录
    threadNum = 20  #下载线程

    databaseDir = '/Users/liurongtao/Documents/Traffic/GoogleWMTS'
    filename = 'huizhou'

    createCacheStruc(extAry, lvRange, cacheDir)       #创建缓存目录结构
    loadTask(tempTask, int(threadNum))                    #下载

    tempTask.join()
    print u'下载进度为:%s'%Download.progress
    print u'下载结束'

    print u'正在转换...'

    time.sleep(2)

    visitDir(cacheDir, databaseDir, filename)


wmts.py
#coding=utf-8
'''
Created on 2012-7-19
将google Map转化为WMST服务
@author: fiftyk
'''
import tornado.ioloop
import tornado.web
import sqlite3, cStringIO, math, os, logging
from PIL import Image
from tornado.web import GZipContentEncoding

environ = os.environ

ENV_HOST = environ.get('HOST', '127.0.0.1')
ENV_PORT = environ.get('PORT', '5555')
ENV_TILE_TYPE = environ.get('TILE_TYPE', 'default')

img = Image.new('RGBA', (256, 256), (128, 0, 0, 0))
tmp = cStringIO.StringIO()
img.save(tmp, 'PNG')
blank = tmp.getvalue()

logging.basicConfig(level=logging.DEBUG, filename='main.log')

class WMTSHandler(tornado.web.RequestHandler):
    TileMatrix = '''<TileMatrix>
                <ows:Identifier>%s</ows:Identifier>
                <ScaleDenominator>%s</ScaleDenominator>
                <TopLeftCorner>-20037508.342787 20037508.342787</TopLeftCorner>
                <TileWidth>256</TileWidth>
                <TileHeight>256</TileHeight>
                <MatrixWidth>%s</MatrixWidth>
                <MatrixHeight>%s</MatrixHeight>
            </TileMatrix>'''
    
    url_pattern = {
        'street': 'http://' + ENV_HOST + ':' + ENV_PORT + '/TMS/%s/%s/%s'
    }
    
    def initialize(self):
        pass

    def get(self):
        #?SERVICE=WMTS&VERSION=1.0.0&REQUEST=GetCapabilities
        #TileRow=1&TileCol=0
        row = self.get_argument('TileRow', -1)
        col = self.get_argument('TileCol', -1)
        level = self.get_argument('TileMatrix', -1)
        layer = self.get_argument('layer', 'street')
        
        if row != -1 and col != -1 and level != -1:
            url = self.url_pattern.get(layer)%(col, row, level.replace('EPSG:3785:', ''))
            #print url
            self.redirect(url)
        else:
            self.set_header('Content-Type', 'text/xml;charset=utf-8')
            TileMatrix_List = ''
        
            for i in range(20):
                level = str(i)
                scale = str(559082264.0287178 / math.pow(2,i))
                width = str(1 << i)
                TileMatrix_List += self.TileMatrix%(level, scale, width, width)
            
            self.xml = open('wmts.xml').read()%TileMatrix_List
            self.write(self.xml.replace('localhost', ENV_HOST).replace('5555', ENV_PORT))

    post = get

GZipContentEncoding.CONTENT_TYPES.add('image/png')

class TMS(tornado.web.RequestHandler):
    def initialize(self, baseDir):
        self.baseDir = baseDir

    def get(self, x, y, z):
        if ENV_TILE_TYPE == 'GOOGLE':
            x = 'C%s' % (hex(int(x))[2:].rjust(8, '0'))
            y = 'R%s' % (hex(int(y))[2:].rjust(8, '0'))
            z = 'L%s' % (z.rjust(2, '0'))
            image_path = os.path.join(self.baseDir, z, y, x) + '.png'
        else:
            image_path = os.path.join(self.baseDir, z, x, y) + '.png'
        self.set_header('Content-Type', 'image/png')
        logging.info(image_path)
        if os.path.exists(image_path):
            inn = cStringIO.StringIO()
            Image.open(image_path).save(inn, 'png')
            self.write(inn.getvalue())
        else:
            self.write(blank)
    
application = tornado.web.Application([
    (r'/wmts', WMTSHandler),
    (r'/TMS/(\w+)/(\w+)/(\w+)', TMS, { 'baseDir' : '/data' }),
], '', None, gzip=True)

application.listen(ENV_PORT)
tornado.ioloop.IOLoop.instance().start()
